### ISSUE TITLE
Web - Scan - Failed to load PDF file uploaded in request money scan

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#27680]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Open an report
2. Request money and open scan tab
3. Upload a pdf
4. Open scan report
5. Click on pdf preview

## Expected Result:
 The pdf should be loaded.

## Actual Result:
The pdf is not loaded correctly.

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [ ] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.71.4
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/93399543/8b37854a-44ac-4ce3-a389-88cb0c1ce209

https://github.com/Expensify/App/assets/93399543/ddacb281-c6e9-478f-8e54-68406c7bd452

**Expensify/Expensify Issue URL:**
**Issue reported by:** @Krishna2323 
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1694901873701159

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by garrettmknight** [comment]
Unnassigning since @sonialiap was first!

**Comment by Talha345** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

When a PDF is used for request money via the scan feature, it is not loaded for preview in the attachment modal via the request money report.

### What is the root cause of that problem?

The `MoneyRequestView` passes an image and thumbnail to `ReportActionItemImage` for preview by calculating the thumbnail and image URL using the `getThumbnailAndImageURIs` helper method below.

https://github.com/Expensify/App/blob/6be8fcac69aa11ff84baaae94d3ef1d495fe5123/src/components/ReportActionItem/MoneyRequestView.js#L92-L95

The `ReportActionItemImage` then displays the thumbnail/image and the image source URL is used to open the attachment.

In case of PDF, the `getThumbnailAndImageURIs` method returns the following response:

`{thumbnail: null, image}` where the image is the URL of just a placeholder image i.e ReceiptGeneric.

https://github.com/Expensify/App/blob/6be8fcac69aa11ff84baaae94d3ef1d495fe5123/src/libs/ReceiptUtils.js#L52

Therefore, when the AttachmentView tries to display the PDF, the following if clause succeeds because the `file.name` has a `.pdf `extension but infact the source is an image URL and therefore the PDF is not loaded.

https://github.com/Expensify/App/blob/6be8fcac69aa11ff84baaae94d3ef1d495fe5123/src/components/Attachments/AttachmentView/index.js#L91

### What changes do you think we should make in order to solve the problem?

We can add the following if clause to `getThumbnailAndImageURIs` helper method.This way for a PDF the thumbnail uses the URL of the placeholder image i.e ReceiptGeneric but the image source contains the actual path of the PDF file and hence it is displayed correctly.

```
    if (fileExtension === 'pdf')
    {
        return {thumbnail: image, image: path};
    }
```


### What alternative solutions did you explore? (Optional)

N/A

### Result:


https://github.com/Expensify/App/assets/36226639/a56f460c-9550-45a9-8bda-fcdd2d71b31b

**Comment by pradeepmdk** [comment]
### Proposal
### Please re-state the problem that we are trying to solve in this issue.
Failed to load PDF file uploaded in request money scan 
### What is the root cause of that problem?
here we are returning the static image even pdf. 

https://github.com/Expensify/App/blob/6be8fcac69aa11ff84baaae94d3ef1d495fe5123/src/libs/ReceiptUtils.js#L64
static image used in preview 
https://github.com/Expensify/App/blob/6be8fcac69aa11ff84baaae94d3ef1d495fe5123/src/components/ReportActionItem/ReportActionItemImage.js#L62

### What changes do you think we should make in order to solve the problem?

we should change this 
https://github.com/Expensify/App/blob/6be8fcac69aa11ff84baaae94d3ef1d495fe5123/src/libs/ReceiptUtils.js#L64
to 
`return {thumbnail: image, image: path};`
so that all static image show in the thumbnail and the original path take to preview 
and when we change this we should change `ThumbnailImage`  component props allow image as reference number

because in native device local import as reference number we already did this some of component
like this 
https://github.com/Expensify/App/blob/6be8fcac69aa11ff84baaae94d3ef1d495fe5123/src/components/ReportActionItem/ReportActionItemImage.js#L18-L19

when we change this it will create regression here 
https://github.com/Expensify/App/blob/6be8fcac69aa11ff84baaae94d3ef1d495fe5123/src/components/MoneyRequestConfirmationList.js#L445
so here we need to change 
```javascript
let receiptPathImage;
    if (!_.isEmpty(props.receiptPath)) {
       const image =  ReceiptUtils.getThumbnailAndImageURIs(props.receiptPath, props.receiptSource)
       receiptPathImage = image.thumbnail || image.image
    }

......
source={{uri: receiptPathImage}}
``` 
here thumbnail is always an image except in this case
```
 if (isReceiptImage && (path.startsWith('blob:') || path.startsWith('file:'))) {
        return {thumbnail: null, image: path};
    }
```


this will work for another type as well like doc, svg

**Comment by sonialiap** [comment]
@sobitneupane looks like we have proposals rolling in, any thoughts on the ones we currently have?

**Comment by sobitneupane** [comment]
Thanks for the proposal @ewanmellor

> We need to pass the proper attachment filepath so that the AttachmentModal can be given the correct info

Your solution should include how you are going to achieve the result.

**Comment by sobitneupane** [comment]
@pradeepmdk Thanks for the proposal.

What change are you suggesting in `ThumbnailImage` and how and what regression will it cause regression? Can we return `{thumbnail: image, image: path};` only if file is pdf?

**Comment by sobitneupane** [comment]
DO we have any issue with docx?

**Comment by Talha345** [comment]
> DO we have any issue with docx?

@sobitneupane With docx, currently we are displaying the docx generic image because we do not have the capability to display docx attachments even when we send them in the chat.


https://github.com/Expensify/App/assets/36226639/22e82145-40ff-42f4-8609-b0ab28bce533

**Comment by pradeepmdk** [comment]
@sobitneupane 

> @pradeepmdk Thanks for the proposal.
> 
> What change are you suggesting in `ThumbnailImage` and how and what regression will it cause regression? Can we return `{thumbnail: image, image: path};` only if file is pdf?

if we need only pdf we can add but docs again we need add one more condition show the preview like this 
![Screenshot 2023-09-25 at 8 33 34 PM](https://github.com/Expensify/App/assets/16595846/26e8045f-17a2-4d92-9503-827a64a9c41d)

and `{thumbnail: image, image: path};` this will create regression so we need to fix that as well.

https://github.com/Expensify/App/assets/16595846/c67a2e1a-5bd2-4bbe-95b4-0b3e5860ba7e

and in `ThumbnailImage`

https://github.com/Expensify/App/blob/f20eba24130d8bfe9e15e9e30c619477711af1f1/src/components/ThumbnailImage.js#L13

here just we need update props allow reference number like this 
`image: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
`

**Comment by sobitneupane** [comment]
@pradeepmdk Thumbnail is infinitely loading on my end. Is it working fine in your end? `getThumbnailAndImageURIs`  is used in multiple places, we must make sure everyone of them are working as expected.

What will be the end result with doc and docx file after the change? Screen Recording will be helpful

https://github.com/Expensify/App/assets/25876548/8554de62-ccab-4af6-9f90-34cd2ce12c4c

**Comment by Talha345** [comment]
@sobitneupane Please confirm if we want to change the behavior for doc and docx files too.

The reason I am suggesting to not change the behavior for these files is because we do not have the capability to display doc files in the attachment preview modal.If we make the change, only the name of the attached file is shown in the attachment modal and the UX result is pretty bad.

**Comment by pradeepmdk** [comment]
@sobitneupane For me working fine.

https://github.com/Expensify/App/assets/16595846/c86aad54-f823-48c8-8201-cc8c7a579cae

> What will be the end result with doc and docx file after the change?

here doc file is converted as a pdf now in requestmoney api.

https://github.com/Expensify/App/assets/16595846/3e943187-d7ab-4a79-b75f-c8dd6ae9ee15


https://github.com/Expensify/App/assets/16595846/2f938ddc-94ed-4cb1-a896-b3c7e24c962f



> getThumbnailAndImageURIs is used in multiple places, we must make sure every one of them is working as expected.

this is covered.
[App/src/components/MoneyRequestConfirmationList.js](https://github.com/Expensify/App/blob/6be8fcac69aa11ff84baaae94d3ef1d495fe5123/src/components/MoneyRequestConfirmationList.js#L445) here only it will create regression.

**Comment by sobitneupane** [comment]
@sonialiap We now allow users to even upload videos as receipts. Which file types should we permit users to upload? 

Currently, the app doesn't support docx preview. I suggest that when users click on a docx file (and other file types), it should trigger a download. Only for PDFs and images, I recommend opening the Attachment Modal to display these files. What are your thoughts or recommendations on this?

**Comment by pradeepmdk** [comment]
@sobitneupane 
here 
https://github.com/Expensify/App/blob/a4efd74857461a13cd66663824317676538a182e/src/pages/iou/ReceiptSelector/index.js#L90
we are declared `UNALLOWED_EXTENSIONS` but no key in the const file 
https://github.com/Expensify/App/blob/a4efd74857461a13cd66663824317676538a182e/src/CONST.ts#L47-L52

> Currently, the app doesn't support docx preview. 

currently, docx is converted to pdf from the backend.

**Comment by Talha345** [comment]
@sobitneupane I tested the flow again with all file types and in case of HTML and DOC/DOCX files, the backend tries to read these files and converts them into a PDF for display.If it succeeds, a PDF is displayed.For files it is unable to read and convert we just show the following view:

![image](https://github.com/Expensify/App/assets/36226639/4f9e01d9-2546-463b-8174-6d145767f2f4)

Also, the method `getThumbnailAndImageURIs` can be optimized and re-written as follows:

```
function getThumbnailAndImageURIs(path: string, filename: string): ThumbnailAndImageURI {
    const isReceiptImage = Str.isImage(filename);
    let thumbnail = null;

    if (isReceiptImage && (path.startsWith('blob:') || path.startsWith('file:'))) {
        return {thumbnail, image: path};
    }

    if (isReceiptImage) {
        thumbnail = `${path}.1024.jpg`;
    } else {
        const {fileExtension} = FileUtils.splitExtensionFromFileName(filename) as FileNameAndExtension;

        if (fileExtension === CONST.IOU.FILE_TYPES.HTML) {
  
```          thumbnail = ReceiptHTML;
        } else if (fileExtension === CONST.IOU.FILE_TYPES.DOC || fileExtension === CONST.IOU.FILE_TYPES.DOCX) {
            thumbnail = ReceiptDoc;
        } else if (fileExtension === CONST.IOU.FILE_TYPES.SVG) {
            thumbnail = ReceiptSVG;
        } else if (fileExtension === CONST.IOU.FILE_TYPES.PDF) {
            thumbnail = ReceiptGeneric;
        }
    }

    return {thumbnail, image: path};
}

**Comment by sobitneupane** [comment]
Regarding [this](https://github.com/Expensify/App/issues/27680#issuecomment-1746163104), I have started [discussion](https://expensify.slack.com/archives/C01GTK53T8Q/p1696936253355669) in slack.

**Comment by sonialiap** [comment]
Thanks Sobit! I'm OOO until Oct 22nd so I won't be around to push the conversation forward until then. Please do so if you don't get traction

**Comment by trjExpensify** [comment]
Posted here in that thread to try and help move this along: https://expensify.slack.com/archives/C01GTK53T8Q/p1697672309044349?thread_ts=1696936253.355669&cid=C01GTK53T8Q

Also CC: @dylanexpensify this should be tracked in the wave4 receipt scan project, so dropping it in there accordingly.

**Comment by sonialiap** [comment]
In the thread it was decided that we'll wait for Sobit to come back from OOO to look at the proposals. Currently his status shows that he's OOO until Oct 27

**Comment by Gonals** [comment]
Waiting for Sobit!

**Comment by dylanexpensify** [comment]
@sobitneupane any chance you're back for this?

**Comment by sobitneupane** [comment]
Sorry for the delay. I will try to get to this issue by tomorrow.

**Comment by trjExpensify** [comment]
Thanks, @sobitneupane. It's pretty high value for our TU campaign so the team can see PDF receipts on NewDot and not have to try and find them on OldDot.

**Comment by sobitneupane** [comment]
Based on the [discussion](https://expensify.slack.com/archives/C01GTK53T8Q/p1696936253355669), it was decided to focus only on displaying pdf in this issue.

**Comment by sobitneupane** [comment]
@bernhardoj 

The expected result here is to display pdf whenever we click on the pdf uploaded as receipt. As of now, we are unable to open such files with "Failed to load PDF file." message.

Does your proposal meet the achieve the expected result?

**Comment by sobitneupane** [comment]
Thanks for the proposal everyone.

[Proposal](https://github.com/Expensify/App/issues/27680#issuecomment-1730711816) from @pradeepmdk looks good to me.

🎀 👀 🎀 C+ reviewed

**Comment by rlinoz** [comment]
> and when we change this we should change ThumbnailImage component props allow image as reference number

I don't really like the idea of making the `ThumbnailImage` component accept something other than a URL since it should be the downsized version of the image.

What if we check if it is a local resource and use the `Image` component instead?

@sobitneupane  @pradeepmdk

**Comment by Talha345** [comment]
> Thanks for the proposal everyone.
> 
> [Proposal](https://github.com/Expensify/App/issues/27680#issuecomment-1730711816) from @pradeepmdk looks good to me.
> 
> 🎀 👀 🎀 C+ reviewed

@sobitneupane I believe my [proposal](https://github.com/Expensify/App/issues/27680#issuecomment-1730489560) was the first one to solve the issue and @pradeepmdk's solution is similar to mine.Why was it not considered?

**Comment by pradeepmdk** [comment]
> > and when we change this we should change ThumbnailImage component props allow image as reference number
> 
> I don't really like the idea of making the `ThumbnailImage` component accept something other than a URL since it should be the downsized version of the image.
> 
> What if we check if it is a local resource and use the `Image` component instead?
> 
> @sobitneupane @pradeepmdk


as per my understanding we don't have a problem here because in ImageRenderer
https://github.com/Expensify/App/blob/main/src/components/HTMLEngineProvider/HTMLRenderers/ImageRenderer.js#L82
we always use this thumbnail image only.

**Comment by pradeepmdk** [comment]
@rlinoz  @sobitneupane  because after modal opened again fetch image from getThumbnailAndImageURIs 
for simple fix for this we can pass 
https://github.com/Expensify/App/blob/main/src/components/Attachments/AttachmentCarousel/extractAttachmentsFromReport.js#L56
to 
```javascript
   const {image, path} = ReceiptUtils.getThumbnailAndImageURIs(transaction);
                const isLocalFile = typeof image === 'string' && (image.startsWith('blob:') || image.startsWith('file:'));
                attachments.unshift({
                    source: tryResolveUrlFromApiRoot(path || image),
                    isAuthTokenRequired: !isLocalFile,
                    file: {name: transaction.filename},
                    isReceipt: true,
                    transactionID,
                });
                return;
``` 
and in getThumbnailAndImageURIs
```javascript
 if (fileExtension === 'pdf')
    {
        return {thumbnail: null, image, path};
    }
```

**Comment by sobitneupane** [comment]
@Talha345 Your proposal was later updated to include details which was not initially present in your proposal as you mentioned [here](https://github.com/Expensify/App/issues/27680#issuecomment-1733343124). So, I went with the first proposal with the required details.

**Comment by Talha345** [comment]
> @Talha345 Your proposal was later updated to include details which was not initially present in your proposal as you mentioned [here](https://github.com/Expensify/App/issues/27680#issuecomment-1733343124). So, I went with the first proposal with the required details.

@sobitneupane The update was regarding fixing a side effect that was caused after applying the proposed solution. The original proposal fixed the issue at hand. Such regressions are often introduced and usually it is not required to cater the regressions in proposals.

For example, this issue is now 1 month old and now while implementing the proposed solution there might be further regressions but that can obviously be discussed and fixed in the PR itself. The guidelines also stated that you don't have to post the entire code in the proposal, just the first proposal that points in the correct direction should be chosen.

**Comment by sobitneupane** [comment]
> I don't really like the idea of making the ThumbnailImage component accept something other than a URL since it should be the downsized version of the image.

> What if we check if it is a local resource and use the Image component instead?

@rlinoz We can return another variable `isLocalFile` from `getThumbnailAndImageURIs()` and use it in `ReportActionItemImage` to determine wether to use `ThumbnailImage` or `Image`.

**Comment by rlinoz** [comment]
Sounds good, let's go with that proposal.

Thanks!

**Comment by pradeepmdk** [comment]
@sonialiap any update for payment

**Comment by sonialiap** [comment]
@pradeepmdk fix = $500 - offer sent
@sobitneupane review = $500 - please request in ND
@Krishna2323 report = $50 - offer sent

**Comment by sonialiap** [comment]
@sobitneupane please complete the [checklist](https://github.com/Expensify/App/issues/27680#issuecomment-1813240758)

**Comment by sobitneupane** [comment]
Sorry for the delay. I will get it done asap.

**Comment by sobitneupane** [comment]
> **BugZero Checklist:** The PR fixing this issue has been merged! The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:
> 
> * [ ]  [@sobitneupane] The PR that introduced the bug has been identified. Link to the PR:

https://github.com/Expensify/App/pull/24235 was the PR which introduced the feature of SmartScan to newDot. The PR mainly focused on images as receipt. So, Rather than a bug, I will say it was work on progress.
> * [ ]  [@sobitneupane] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:

N/A
> * [ ]  [@sobitneupane] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:

N/A
> * [ ]  [@sobitneupane] Determine if we should create a regression test for this bug.

Yes.
> * [ ]  [@sobitneupane] If we decide to create a regression test for the bug, please propose the [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps to ensure the same bug will not reach production again.

**Comment by sobitneupane** [comment]
Regression Test Proposal:

1. Request money and open scan tab
2. Upload a pdf
3. Open scanned report
4. Click on pdf preview
5. Verify that pdf is loaded.

Do we agree 👍 or 👎

**Comment by sonialiap** [comment]
Thanks, Sobit! Submitting regression test and closing this out

**Comment by sobitneupane** [comment]
https://github.com/Expensify/App/issues/27680#issuecomment-1827425084

Requested payment on newDot.

**Comment by JmillsExpensify** [comment]
$500 payment approved for @sobitneupane based on [this comment](https://github.com/Expensify/App/issues/27680#issuecomment-1827425084).

**Comment by trjExpensify** [comment]
👋 Heads up! @pradeepmdk @sobitneupane @Gonals this seems to have broken again:

<img width="1172" alt="image" src="https://github.com/Expensify/App/assets/16232057/2651891f-e5a4-4170-92bb-f8a304c7f177">

<img width="987" alt="image" src="https://github.com/Expensify/App/assets/16232057/0a878050-e15f-4255-8e62-8985f901a3a8">

Any chance you'd be willing to help us diagnose so we can create a follow-up issue/PR to fix it, or maybe encapsulate it into this issue? 

CC: @RachCHopkins

**Comment by sobitneupane** [comment]
I found the issue. 

https://github.com/Expensify/App/blob/880254c55ea2c210f11d77974cc29e4c446f47bd/src/components/ReportActionItem/ReportActionItemImage.js#L87

`isLocalFile` is true even in the case when authToken is required (making `isAuthTokenRequired` false for even non-local files). We didn't face the issue before because we weren't using `AttachmentModal` component in `ReportActionItemImage`.

@pradeepmdk Will you be willing to create PR to fix the issue.

**Comment by trjExpensify** [comment]
Niceee sleuthing! ❤️ @pradeepmdk could you let us know today if you'd be up for fixing that?

**Comment by Talha345** [comment]
@trjExpensify @sobitneupane  I would be available to fix this follow up issue if @pradeepmdk is not available anymore as I was also involved in this during the proposal and discussion phase! 😊

**Comment by trjExpensify** [comment]
Nice, let's go for it! Fresh issue here: https://github.com/Expensify/App/issues/33412
